File: Iced/Intel/BlockEncoder.cs
/*
Copyright (C) 2018-2019 de4dot@gmail.com

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#if !NO_ENCODER
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Iced.Intel.BlockEncoderInternal;

namespace Iced.Intel {
	/// <summary>
	/// Relocation kind
	/// </summary>
	public enum RelocKind {
		/// <summary>
		/// 64-bit offset. Only used if it's 64-bit code.
		/// </summary>
		Offset64,
	}

	/// <summary>
	/// Relocation info
	/// </summary>
	public readonly struct RelocInfo {
		/// <summary>
		/// Address
		/// </summary>
		public readonly ulong Address;

		/// <summary>
		/// Relocation kind
		/// </summary>
		public readonly RelocKind Kind;

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="kind">Relocation kind</param>
		/// <param name="address">Address</param>
		public RelocInfo(RelocKind kind, ulong address) {
			Kind = kind;
			Address = address;
		}
	}

	/// <summary>
	/// Contains a list of instructions and a base IP
	/// </summary>
	public readonly struct InstructionBlock {
		/// <summary>
		/// Code writer
		/// </summary>
		public readonly CodeWriter CodeWriter;

		/// <summary>
		/// All instructions
		/// </summary>
		public readonly IList<Instruction> Instructions;

		/// <summary>
		/// Base IP of all encoded instructions
		/// </summary>
		public readonly ulong RIP;

		/// <summary>
		/// List that gets all reloc infos or null
		/// </summary>
		public readonly IList<RelocInfo> RelocInfos;

		/// <summary>
		/// Offsets of the new instructions relative to the base IP. If the instruction was rewritten to
		/// a completely different instruction, the value <see cref="uint.MaxValue"/> is stored in that array element.
		/// This array can be null.
		/// </summary>
		public readonly uint[] NewInstructionOffsets;

		/// <summary>
		/// Offsets of all constants in the new encoded instructions. If the instruction was rewritten,
		/// the 'default' value is stored in the corresponding array element. This array can be null.
		/// </summary>
		public readonly ConstantOffsets[] ConstantOffsets;

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="codeWriter">Code writer</param>
		/// <param name="instructions">Instructions</param>
		/// <param name="rip">Base IP of all encoded instructions</param>
		/// <param name="relocInfos">List that gets all reloc infos or null</param>
		/// <param name="newInstructionOffsets">Offsets of the new instructions relative to the base IP. If the instruction was rewritten to
		/// a completely different instruction, the value <see cref="uint.MaxValue"/> is stored in that array element.
		/// This array can be null.</param>
		/// <param name="constantOffsets">Offsets of all constants in the new encoded instructions. If the instruction was rewritten,
		/// the 'default' value is stored in the corresponding array element. This array can be null.</param>
		public InstructionBlock(CodeWriter codeWriter, IList<Instruction> instructions, ulong rip, IList<RelocInfo> relocInfos = null, uint[] newInstructionOffsets = null, ConstantOffsets[] constantOffsets = null) {
			CodeWriter = codeWriter ?? throw new ArgumentNullException(nameof(codeWriter));
			Instructions = instructions ?? throw new ArgumentNullException(nameof(instructions));
			RIP = rip;
			RelocInfos = relocInfos;
			NewInstructionOffsets = newInstructionOffsets;
			ConstantOffsets = constantOffsets;
			if (newInstructionOffsets != null && newInstructionOffsets.Length != instructions.Count)
				throw new ArgumentException($"{nameof(newInstructionOffsets)} must have the same number of elements as {nameof(instructions)} or it must be null");
			if (constantOffsets != null && constantOffsets.Length != instructions.Count)
				throw new ArgumentException($"{nameof(constantOffsets)} must have the same number of elements as {nameof(instructions)} or it must be null");
		}
	}

	/// <summary>
	/// Encoder options
	/// </summary>
	[Flags]
	public enum BlockEncoderOptions : uint {
		/// <summary>
		/// No option is set
		/// </summary>
		None						= 0,

		/// <summary>
		/// By default, branches get updated if the target is too far away, eg. jcc short -> jcc near
		/// or if 64-bit mode, jcc + jmp [rip+mem]. If this option is enabled, no branches are fixed.
		/// </summary>
		DontFixBranches				= 0x00000001,
	}

	/// <summary>
	/// Encodes instructions
	/// </summary>
	public sealed class BlockEncoder {
		readonly int bitness;
		readonly BlockEncoderOptions options;
		readonly Block[] blocks;
		readonly Encoder nullEncoder;
		readonly Dictionary<ulong, Instr> toInstr;

		internal int Bitness => bitness;
		internal bool FixBranches => (options & BlockEncoderOptions.DontFixBranches) == 0;
		internal Encoder NullEncoder => nullEncoder;

		sealed class NullCodeWriter : CodeWriter {
			public static readonly NullCodeWriter Instance = new NullCodeWriter();
			NullCodeWriter() { }
			public override void WriteByte(byte value) { }
		}

		BlockEncoder(int bitness, InstructionBlock[] instrBlocks, BlockEncoderOptions options) {
			if (bitness != 16 && bitness != 32 && bitness != 64)
				throw new ArgumentOutOfRangeException(nameof(bitness));
			if (instrBlocks == null)
				throw new ArgumentNullException(nameof(instrBlocks));
			this.bitness = bitness;
			nullEncoder = Encoder.Create(bitness, NullCodeWriter.Instance);
			this.options = options;

			blocks = new Block[instrBlocks.Length];
			int instrCount = 0;
			for (int i = 0; i < instrBlocks.Length; i++) {
				var instructions = instrBlocks[i].Instructions;
				if (instructions == null)
					throw new ArgumentException();
				var instrs = new Instr[instructions.Count];
				ulong ip = instrBlocks[i].RIP;
				for (int j = 0; j < instrs.Length; j++) {
					var instruction = instructions[j];
					var instr = Instr.Create(this, ref instruction);
					instr.IP = ip;
					instrs[j] = instr;
					instrCount++;
					ip += instr.Size;
				}
				blocks[i] = new Block(this, instrBlocks[i].CodeWriter, instrBlocks[i].RIP, instrBlocks[i].RelocInfos, instrBlocks[i].NewInstructionOffsets, instrBlocks[i].ConstantOffsets, instrs);
			}
			// Optimize from low to high addresses
			Array.Sort(blocks, (a, b) => a.RIP.CompareTo(b.RIP));

			// There must not be any instructions with the same IP, except if IP = 0 (default value)
			var toInstr = new Dictionary<ulong, Instr>(instrCount);
			this.toInstr = toInstr;
			bool hasMultipleZeroIPInstrs = false;
			foreach (var block in blocks) {
				foreach (var instr in block.Instructions) {
					ulong origIP = instr.OrigIP;
					if (toInstr.TryGetValue(origIP, out var origInstr)) {
						if (origIP != 0)
							throw new ArgumentException($"Multiple instructions with the same IP: 0x{origIP:X}");
						hasMultipleZeroIPInstrs = true;
					}
					else
						toInstr[origIP] = instr;
				}
			}
			if (hasMultipleZeroIPInstrs)
				toInstr.Remove(0);

			foreach (var block in blocks) {
				ulong ip = block.RIP;
				foreach (var instr in block.Instructions) {
					instr.IP = ip;
					var oldSize = instr.Size;
					instr.Initialize();
					if (instr.Size > oldSize)
						throw new InvalidOperationException();
					ip += instr.Size;
				}
			}
		}

		/// <summary>
		/// Encodes instructions. Returns null or an error message
		/// </summary>
		/// <param name="bitness">16, 32 or 64</param>
		/// <param name="block">All instructions</param>
		/// <param name="errorMessage">Updated with an error message if the method failed</param>
		/// <param name="options">Encoder options</param>
		/// <returns></returns>
		public static bool TryEncode(int bitness, InstructionBlock block, out string errorMessage, BlockEncoderOptions options = BlockEncoderOptions.None) =>
			TryEncode(bitness, new[] { block }, out errorMessage, options);

		/// <summary>
		/// Encodes instructions. Returns null or an error message
		/// </summary>
		/// <param name="bitness">16, 32 or 64</param>
		/// <param name="blocks">All instructions</param>
		/// <param name="errorMessage">Updated with an error message if the method failed</param>
		/// <param name="options">Encoder options</param>
		/// <returns></returns>
		public static bool TryEncode(int bitness, InstructionBlock[] blocks, out string errorMessage, BlockEncoderOptions options = BlockEncoderOptions.None) =>
			new BlockEncoder(bitness, blocks, options).Encode(out errorMessage);

		bool Encode(out string errorMessage) {
			const int MAX_ITERS = 1000;
			for (int iter = 0; iter < MAX_ITERS; iter++) {
				bool updated = false;
				foreach (var block in blocks) {
					ulong ip = block.RIP;
					foreach (var instr in block.Instructions) {
						instr.IP = ip;
						var oldSize = instr.Size;
						if (instr.Optimize()) {
							if (instr.Size > oldSize) {
								errorMessage = "Internal error: new size > old size";
								return false;
							}
							if (instr.Size < oldSize)
								updated = true;
						}
						else if (instr.Size != oldSize) {
							errorMessage = "Internal error: new size != old size";
							return false;
						}
						ip += instr.Size;
					}
				}
				if (!updated)
					break;
			}

			foreach (var block in blocks)
				block.InitializeData();

			foreach (var block in blocks) {
				var encoder = Encoder.Create(bitness, block.CodeWriter);
				ulong ip = block.RIP;
				var newInstructionOffsets = block.NewInstructionOffsets;
				var constantOffsets = block.ConstantOffsets;
				var instructions = block.Instructions;
				for (int i = 0; i < instructions.Length; i++) {
					var instr = instructions[i];
					uint bytesWritten = block.CodeWriter.BytesWritten;
					bool isOriginalInstruction;
					if (constantOffsets != null)
						errorMessage = instr.TryEncode(encoder, out constantOffsets[i], out isOriginalInstruction);
					else
						errorMessage = instr.TryEncode(encoder, out _, out isOriginalInstruction);
					if (errorMessage != null)
						return false;
					uint size = block.CodeWriter.BytesWritten - bytesWritten;
					if (size != instr.Size) {
						errorMessage = "Internal error: didn't write all bytes";
						return false;
					}
					if (newInstructionOffsets != null) {
						if (isOriginalInstruction)
							newInstructionOffsets[i] = (uint)(ip - block.RIP);
						else
							newInstructionOffsets[i] = uint.MaxValue;
					}
					ip += size;
				}
				block.WriteData();
			}

			errorMessage = null;
			return true;
		}

		internal TargetInstr GetTarget(ulong address) {
			if (toInstr.TryGetValue(address, out var instr))
				return new TargetInstr(instr);
			return new TargetInstr(address);
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/enc/BlockEncoder.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.enc;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import com.github.icedland.iced.x86.Code;
import com.github.icedland.iced.x86.CodeWriter;
import com.github.icedland.iced.x86.ConstantOffsets;
import com.github.icedland.iced.x86.Instruction;
import com.github.icedland.iced.x86.internal.IcedConstants;

/**
 * Encodes instructions. It can be used to move instructions from one location to another location.
 */
public final class BlockEncoder {
	private final int bitness;
	private final int options;
	private final Block[] blocks;
	private final Encoder nullEncoder;
	private final HashMap<Long, Instr> toInstr;

	int getBitness() {
		return bitness;
	}

	boolean fixBranches() {
		return (options & BlockEncoderOptions.DONT_FIX_BRANCHES) == 0;
	}

	private boolean returnRelocInfos() {
		return (options & BlockEncoderOptions.RETURN_RELOC_INFOS) != 0;
	}

	private boolean returnNewInstructionOffsets() {
		return (options & BlockEncoderOptions.RETURN_NEW_INSTRUCTION_OFFSETS) != 0;
	}

	private boolean returnConstantOffsets() {
		return (options & BlockEncoderOptions.RETURN_CONSTANT_OFFSETS) != 0;
	}

	private boolean returnAllConstantOffsets() {
		return (options & BlockEncoderOptions.RETURN_ALL_NEW_INSTRUCTION_OFFSETS) != 0;
	}

	private static final class NullCodeWriter implements CodeWriter {
		public static final NullCodeWriter instance = new NullCodeWriter();

		NullCodeWriter() {
		}

		@Override
		public void writeByte(byte value) {
		}
	}

	BlockEncoder(int bitness, InstructionBlock[] instrBlocks, int options) {
		if (bitness != 16 && bitness != 32 && bitness != 64)
			throw new IllegalArgumentException("bitness");
		if (instrBlocks == null)
			throw new NullPointerException("instrBlocks");
		this.bitness = bitness;
		nullEncoder = new Encoder(bitness, NullCodeWriter.instance);
		this.options = options;

		blocks = new Block[instrBlocks.length];
		int instrCount = 0;
		for (int i = 0; i < instrBlocks.length; i++) {
			List<Instruction> instructions = instrBlocks[i].instructions;
			if (instructions == null)
				throw new IllegalArgumentException();
			Block block = new Block(this, instrBlocks[i].codeWriter, instrBlocks[i].rip, returnRelocInfos() ? new ArrayList<RelocInfo>() : null);
			blocks[i] = block;
			Instr[] instrs = new Instr[instructions.size()];
			long ip = instrBlocks[i].rip;
			for (int j = 0; j < instrs.length; j++) {
				Instruction instruction = instructions.get(j);
				Instr instr = Instr.create(this, block, instruction);
				instr.ip = ip;
				instrs[j] = instr;
				instrCount++;
				assert instr.size != 0 || instruction.getCode() == Code.ZERO_BYTES;
				ip += instr.size;
			}
			block.setInstructions(instrs);
		}
		// Optimize from low to high addresses
		Arrays.sort(blocks, (a, b) -> Long.compareUnsigned(a.rip, b.rip));

		// There must not be any instructions with the same IP, except if IP = 0 (default value)
		HashMap<Long, Instr> toInstr = new HashMap<Long, Instr>(instrCount);
		this.toInstr = toInstr;
		boolean hasMultipleZeroIPInstrs = false;
		for (Block block : blocks) {
			for (Instr instr : block.getInstructions()) {
				long origIP = instr.origIP;
				if (toInstr.containsKey(origIP)) {
					if (origIP != 0)
						throw new IllegalArgumentException(String.format("Multiple instructions with the same IP: 0x%X", origIP));
					hasMultipleZeroIPInstrs = true;
				}
				else
					toInstr.put(origIP, instr);
			}
		}
		if (hasMultipleZeroIPInstrs)
			toInstr.remove(0L);

		for (Block block : blocks) {
			long ip = block.rip;
			for (Instr instr : block.getInstructions()) {
				instr.ip = ip;
				if (!instr.done)
					instr.initialize(this);
				ip += instr.size;
			}
		}
	}

	/**
	 * Encodes instructions. Any number of branches can be part of this block.
	 * <p>
	 * You can use this function to move instructions from one location to another location.
	 * <p>
	 * If the target of a branch is too far away, it'll be rewritten to a longer branch.
	 * You can disable this by passing in {@link BlockEncoderOptions#DONT_FIX_BRANCHES}.
	 * <p>
	 * If the block has any {@code RIP}-relative memory operands, make sure the data isn't too
	 * far away from the new location of the encoded instructions. Every OS should have
	 * some API to allocate memory close (+/-2GB) to the original code location.
	 *
	 * @param bitness 16, 32 or 64
	 * @param block   All instructions
	 * @return Object On success, it returns a {@link BlockEncoderResult}, and on failure, it returns a {@link String} with the error message.
	 */
	public static Object tryEncode(int bitness, InstructionBlock block) {
		return tryEncode(bitness, block, BlockEncoderOptions.NONE);
	}

	/**
	 * Encodes instructions. Any number of branches can be part of this block.
	 * <p>
	 * You can use this function to move instructions from one location to another location.
	 * <p>
	 * If the target of a branch is too far away, it'll be rewritten to a longer branch.
	 * You can disable this by passing in {@link BlockEncoderOptions#DONT_FIX_BRANCHES}.
	 * <p>
	 * If the block has any {@code RIP}-relative memory operands, make sure the data isn't too
	 * far away from the new location of the encoded instructions. Every OS should have
	 * some API to allocate memory close (+/-2GB) to the original code location.
	 *
	 * @param bitness 16, 32 or 64
	 * @param block   All instructions
	 * @param options Encoder options (a {@link BlockEncoderOptions} flags value)
	 * @return Object On success, it returns a {@link BlockEncoderResult}, and on failure, it returns a {@link String} with the error message.
	 */
	public static Object tryEncode(int bitness, InstructionBlock block, int options) {
		Object result = tryEncode(bitness, new InstructionBlock[] { block }, options);
		if (result instanceof BlockEncoderResult[]) {
			BlockEncoderResult[] resultArray = (BlockEncoderResult[])result;
			assert resultArray.length == 1 : resultArray.length;
			return resultArray[0];
		}
		if (result instanceof String)
			return (String)result;
		throw new UnsupportedOperationException();
	}

	/**
	 * Encodes instructions. Any number of branches can be part of this block.
	 * <p>
	 * You can use this function to move instructions from one location to another location.
	 * <p>
	 * If the target of a branch is too far away, it'll be rewritten to a longer branch.
	 * You can disable this by passing in {@link BlockEncoderOptions#DONT_FIX_BRANCHES}.
	 * <p>
	 * If the block has any {@code RIP}-relative memory operands, make sure the data isn't too
	 * far away from the new location of the encoded instructions. Every OS should have
	 * some API to allocate memory close (+/-2GB) to the original code location.
	 *
	 * @param bitness 16, 32 or 64
	 * @param blocks  All instructions
	 * @return Object On success, it returns an array of {@link BlockEncoderResult}, and on failure, it returns a {@link String} with the error
	 *         message.
	 */
	public static Object tryEncode(int bitness, InstructionBlock[] blocks) {
		return tryEncode(bitness, blocks, BlockEncoderOptions.NONE);
	}

	/**
	 * Encodes instructions. Any number of branches can be part of this block.
	 * <p>
	 * You can use this function to move instructions from one location to another location.
	 * <p>
	 * If the target of a branch is too far away, it'll be rewritten to a longer branch.
	 * You can disable this by passing in {@link BlockEncoderOptions#DONT_FIX_BRANCHES}.
	 * <p>
	 * If the block has any {@code RIP}-relative memory operands, make sure the data isn't too
	 * far away from the new location of the encoded instructions. Every OS should have
	 * some API to allocate memory close (+/-2GB) to the original code location.
	 *
	 * @param bitness 16, 32 or 64
	 * @param blocks  All instructions
	 * @param options Encoder options (a {@link BlockEncoderOptions} flags value)
	 * @return Object On success, it returns an array of {@link BlockEncoderResult}, and on failure, it returns a {@link String} with the error
	 *         message.
	 */
	public static Object tryEncode(int bitness, InstructionBlock[] blocks, int options) {
		return new BlockEncoder(bitness, blocks, options).encode();
	}

	private Object encode() {
		final int MAX_ITERS = 5;
		for (int iter = 0; iter < MAX_ITERS; iter++) {
			boolean updated = false;
			for (Block block : blocks) {
				long ip = block.rip;
				long gained = 0;
				for (Instr instr : block.getInstructions()) {
					instr.ip = ip;
					if (!instr.done) {
						int oldSize = instr.size;
						if (instr.optimize(gained)) {
							if (instr.size > oldSize)
								return "Internal error: new size > old size";
							if (instr.size < oldSize) {
								gained += oldSize - instr.size;
								updated = true;
							}
						}
						else if (instr.size != oldSize)
							return "Internal error: new size != old size";
					}
					ip += instr.size;
				}
			}
			if (!updated)
				break;
		}

		for (Block block : blocks)
			block.initializeData();

		BlockEncoderResult[] resultArray = new BlockEncoderResult[blocks.length];
		TryEncodeResult tryEncResult = new TryEncodeResult();
		boolean returnAllOffsets = returnAllConstantOffsets();
		for (int i = 0; i < blocks.length; i++) {
			Block block = blocks[i];
			Encoder encoder = new Encoder(bitness, block.codeWriter);
			long ip = block.rip;
			int[] newInstructionOffsets = returnNewInstructionOffsets() ? new int[block.getInstructions().length] : null;
			ConstantOffsets[] constantOffsets = returnConstantOffsets() ? new ConstantOffsets[block.getInstructions().length] : null;
			Instr[] instructions = block.getInstructions();
			for (int j = 0; j < instructions.length; j++) {
				Instr instr = instructions[j];
				int bytesWritten = block.codeWriter.bytesWritten;
				String errorMessage = instr.tryEncode(encoder, tryEncResult);
				if (errorMessage != null)
					return errorMessage;
				if (constantOffsets != null)
					constantOffsets[j] = tryEncResult.constantOffsets;
				int size = block.codeWriter.bytesWritten - bytesWritten;
				if (size != instr.size)
					return "Internal error: didn't write all bytes";
				if (newInstructionOffsets != null) {
					if (tryEncResult.isOriginalInstruction || returnAllOffsets)
						newInstructionOffsets[j] = (int)(ip - block.rip);
					else
						newInstructionOffsets[j] = 0xFFFF_FFFF;
				}
				ip += size;
			}
			resultArray[i] = new BlockEncoderResult(block.rip, block.relocInfos, newInstructionOffsets, constantOffsets);
			block.writeData();
		}

		return resultArray;
	}

	TargetInstr getTarget(long address) {
		Instr instr = toInstr.get(address);
		if (instr != null)
			return new TargetInstr(instr);
		return new TargetInstr(address);
	}

	int getInstructionSize(Instruction instruction, long ip) {
		Object result = nullEncoder.tryEncode(instruction, ip);
		if (result instanceof Integer)
			return ((Integer)result).intValue();
		return IcedConstants.MAX_INSTRUCTION_LENGTH;
	}
}
